File: remote/src/main/java/com/kelsos/mbrc/presenters/LyricsPresenterImpl.java
package com.kelsos.mbrc.presenters;

import com.google.inject.Inject;
import com.kelsos.mbrc.events.ui.LyricsChangedEvent;
import com.kelsos.mbrc.interactors.TrackLyricsInteractor;
import com.kelsos.mbrc.ui.views.LyricsView;
import com.kelsos.mbrc.utilities.RxBus;
import roboguice.util.Ln;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

public class LyricsPresenterImpl implements LyricsPresenter {

  private LyricsView view;
  @Inject private RxBus bus;
  @Inject private TrackLyricsInteractor lyricsInteractor;

  @Override public void bind(LyricsView view) {
    this.view = view;
    bus.register(this, LyricsChangedEvent.class, this::updateLyricsData);
  }

  @Override public void onPause() {

  }

  @Override public void onResume() {
    loadLyrics();
  }

  public void updateLyricsData(LyricsChangedEvent update) {
    view.updateLyrics(update.getLyrics());
  }

  private void loadLyrics() {
    lyricsInteractor.execute(false)
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(view::updateLyrics, Ln::v);
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/presenters/LyricsPresenterImpl.kt
package com.kelsos.mbrc.presenters

import com.kelsos.mbrc.events.ui.LyricsChangedEvent
import com.kelsos.mbrc.extensions.task
import com.kelsos.mbrc.interactors.TrackLyricsInteractor
import com.kelsos.mbrc.ui.views.LyricsView
import com.kelsos.mbrc.utilities.RxBus
import timber.log.Timber
import javax.inject.Inject

class LyricsPresenterImpl
@Inject constructor(private val bus: RxBus,
                    private val lyricsInteractor: TrackLyricsInteractor) : LyricsPresenter {

  private var view: LyricsView? = null

  override fun bind(view: LyricsView) {
    this.view = view
    bus.register(this, LyricsChangedEvent::class.java, { this.updateLyricsData(it) })
  }

  override fun onPause() {

  }

  override fun onResume() {
    loadLyrics()
  }

  fun updateLyricsData(update: LyricsChangedEvent) {
    view?.updateLyrics(update.lyrics)
  }

  private fun loadLyrics() {
    lyricsInteractor.execute(false).task().subscribe(
        { view?.updateLyrics(it) },
        { Timber.v(it) })
  }
}
